taichi/backends/opengl/opengl_api.cpp

```diff
@@ -285,26 +287,106 @@ struct GLBuffer : GLSSBO {
   }
 };
 
+struct GLBufferAllocator {
+ private:
+  const size_t max_free_resident_size = 64 << 20;
+
+  std::list<std::unique_ptr<GLBuffer>> buffers_storage;
+
+  std::map<std::pair<GLBufId, size_t>, GLBuffer *> buffers_mapping;
```

Comment:
can we use unordered_map instead?

Reply:
Hmm, I guess it's because `unordered_map` can't really hash the key of type `std::pair<GlBufId, size_t>`. I think we can have a very small struct here, both for being the key type (with its hasher), and for making the semantics more clear (i.e. we don't have to write  `first.second` and guess what `second` is...)

Reply:
Actually reading the code a bit I think index is not even needed. Since all buffers are created with essentially the same access hint, and GL does not have the concept of memory pool types (i think), we only need to have the size to be the same to reuse.

Reply:
In Vulkan we will need to only reuse buffers from the same pool / type of pool, but not in GL.

Reply:
An update to this reuse topic: Since I switched to persistent storage (glShaderStorage), there is (kind of) memory pool in GL. For example, a buffer created without the MAP_WRITE_BIT can not be mapped and written to by the host.